pkgs/applications/networking/mailreaders/himalaya/default.nix

```diff
@@ -34,12 +36,17 @@ rustPlatform.buildRustPackage rec {
       openssl
     ];
 
-  # flag added because without end-to-end testing is ran which requires
-  # additional tooling and servers to test
+  # TODO: is there a way to spawn an IMAP server to run integration
+  # tests?
```

Comment:
Hi @soywod,
Nice to meet you here. :wave: 

It could be good to cover the integration test for the app in Rust.
Rust By Example also suggests a way to set up an integration test.
https://doc.rust-lang.org/rust-by-example/testing/integration_testing.html

If you think, the integration test will not really run every time and need to run dependent on the environment smartly without flags.  `test-with` could be a solution for you.
https://github.com/yanganto/test-with

Here is my suggestion to do this, and happy to help you set up the integration test.
An IMAP server executable binary could be set in `nativeBuildInputs` for nix,  such that the test in Rust can now setup it up and do integration tests.  And use [executable condition](https://github.com/yanganto/test-with#executable-condition) of `test-with`, such that we don't need a cargoTestFlag, and other contributors will not worry to run test cases with the corresponding IMAP server on their environment.

Reply:
`nativeBuildInputs` or `nativeCheckInputs`??

Reply:
@toastal Good catch,  thanks.

Reply:
Hi @soywod  
Here is some information could be helpful.

`openjdk17-bootstrap` and `greenmail` work well with this command 

`java -Dgreenmail.setup.test.all -Dgreenmail.hostname=0.0.0.0 -Dgreenmail.auth.disabled -Dgreenmail.verbose -jar /tmp/greenmail-standalone-1.6.13.jar`

Reply:
@yanganto nice to meet you back there :slightly_smiling_face: I already have integration tests working for the tool, and I also know how to launch them even in a CI (I mostly use this command: `docker run -it --rm -p 3025:3025 -p 3110:3110 -p 3143:3143 -p 3465:3465 -p 3993:3993 -p 3995:3995 -e GREENMAIL_OPTS='-Dgreenmail.setup.test.all -Dgreenmail.hostname=0.0.0.0 -Dgreenmail.auth.disabled -Dgreenmail.verbose' greenmail/standalone:1.6.11`). My question was more: how to start this server during nix tests?

> An IMAP server executable binary could be set in `nativeCheckInputs` for nix (It is a dependency for `cargo test`), such that the test in Rust can now setup it up and do integration tests when the nix package is building.
> […]
> `openjdk17-bootstrap` and `greenmail` work well with this command

Sounds great! But I did not understand *when* to launch the IMAP server?

Reply:
Thank you for all those details, but I still don't get *how* and *when* to spawn the `.jar` :sweat_smile: the simplest for everyone would be to use Rust crates instead of an external `.jar`, this way you control when to start / stop it. The best project I found so far is [Stalwart](https://stalw.art/), but they expose only a CLI not the lib (I just asked [the question](https://github.com/stalwartlabs/jmap-server/discussions/31), let's see). Anyway sounds a bit complicated for now, maybe we could leave it for a next PR?

Reply:
These might be a clear response for you.

```text
When -> the check phase of the flow to build the nix package, but handled by `buildRustPackage` for Rust app.
How  ->  1. need to customize `buildRustPackage` and 
         2. docker image(not reasonable) or `.jar` should be included in this commit of nixpkgs.  
  or
  
How  ->  1. control the integration test flow in Rust test case,
         2. include `.jar` in Rust project or the commit here
         3. add `nativeCheckInputs = [openjdk17-bootstrap];`
```

Currently, this PR is good, and this thread is just a comment/suggestion, not a required review.  :smile: And I am not sure it is also good for @toastal.

I have no permission to merge this, the thing I only have is to reject and close this PR.
Disabled some tests not work with Nix flow in the current PR looks good to me,
so I just provided the information about how to make a compatible test flow to make the integration test works with NixOS. 

Nix is a robust system, Rust is a robust language, and we all love these.:heart:
Somehow we need a new thinking for these, using a docker image to test an app is not suitable, there is nowhere to put `docker run ...` and make everything works.  I hope you can understand, I just want to use my experience and try to save you time without pain.  Thanks for your time to read a lot of words :pray:

Reply:
Thank you again for your replies, I understand better now. I just created [an issue](https://todo.sr.ht/~soywod/pimalaya/36) to run integration tests without docker images, which will fix issues for everyone (nix included).